pkg/controller/virtualservice.go

```diff
@@ -284,13 +284,76 @@ func getVServiceCondition(conditionType appmeshv1beta1.VirtualServiceConditionTy
 	return appmeshv1beta1.VirtualServiceCondition{}
 }
 
-func getVirtualRouter(vservice *appmeshv1beta1.VirtualService) *appmeshv1beta1.VirtualRouter {
+func (c *Controller) getVirtualRouter(vservice *appmeshv1beta1.VirtualService) *appmeshv1beta1.VirtualRouter {
+	var vrouter *appmeshv1beta1.VirtualRouter
 	if vservice.Spec.VirtualRouter != nil {
-		return vservice.Spec.VirtualRouter
+		vrouter = vservice.Spec.VirtualRouter
+	} else {
+		vrouter = &appmeshv1beta1.VirtualRouter{
+			Name: vservice.Name,
+		}
+	}
+
+	if len(vrouter.Listeners) == 0 {
+		listener := c.getListenerFromRouteTarget(vservice, vrouter)
+		if listener != nil {
+			vrouter.Listeners = append(vrouter.Listeners, appmeshv1beta1.VirtualRouterListener{
+				PortMapping: listener.PortMapping,
+			})
+		}
+	}
+
+	return vrouter
+}
+
+//getListenerFromRouteTarget populates listener for virtual-router if one is missing.
+//Reason: VirtualRouter requires exactly one listener. Initial versions of CRD did not
+//have listener field defined in virtual-router object. Here we will peek into the
+//first route defined for the virtual-service and use one of the virtual-nodes listener
+//to populate virtual-router listener. In some edge cases this can be error-prone and
+//it is recommended to explicitly define virtual-router.
+//See https://docs.aws.amazon.com/app-mesh/latest/userguide/virtual_routers.html for more information.
+func (c *Controller) getListenerFromRouteTarget(vservice *appmeshv1beta1.VirtualService, vrouter *appmeshv1beta1.VirtualRouter) *appmeshv1beta1.Listener {
+	if len(vservice.Spec.Routes) == 0 {
+		return nil
 	}
-	return &appmeshv1beta1.VirtualRouter{
-		Name: vservice.Name,
+
+	var candidateVnodeName string
+	route := vservice.Spec.Routes[0]
+	if route.Http != nil {
+		if len(route.Http.Action.WeightedTargets) > 0 {
+			candidateVnodeName = route.Http.Action.WeightedTargets[0].VirtualNodeName
```

Comment:
yeah given that it is edge case we can probably error out if virtual-nodes are using different port-mapping. That said it can be deep check that can be inefficient. Only sane thing is to slowly migrate to defining listener on virtual-router.

Reply:
Yeah, my main concern is if someone changes the order of the Virtual Nodes in the route, and those nodes have a different listener port, they'll have broken the downstream clients, potentially without knowing.

Reply:
alternately #74 may be enough to do.

Reply:
Sounds good. Thanks!